### PR TITLE
feat(project-engine-client): model 1-level nested AIO tags in the mock

### PR DESCRIPTION
**Pull request by rainer-friederich**
## 1. Abstract
Teaches the `@adobe/spacecat-shared-project-engine-client` Counterfact mock and its baked seeds to model 1-level nested AIO tags (parent/child via `parent_id`) — parent-aware create, tree-aware listing, a re-parent/rename endpoint, a nested baked fixture — and reconciles the mock + spec overlay to what a live prod probe (2026-07-01) confirmed.

## 2. Reasoning
The client's generated types already model tag nesting (`parent_id`, `path[]`, `children_count`), but the mock was flat and ignored `parent_id`, so nothing local could validate the nested-category feature. Consumers that write `parent_id` (the nested-category work in spacecat-api-service and the onboarding CLI) would pass against the flat mock while silently dropping hierarchy. This closes that gap and, after a live probe, corrects the places where the vendored spec/mock diverged from prod.

## 3. High-level overview of the changes
Behaviour delta in the mock's project-level AIO tag taxonomy (the Categories surface):

- `POST /aio/tags` honours `body.parent_id` (children stored under that category).
- `GET /aio/tags` is tree-aware: `parent_id=''` returns roots, a non-empty `parent_id` returns that category's children; each tag carries a derived `children_count` and, for a child, a single-leaf `path[]` breadcrumb (its root). A flat root reads back `parent_id: null` and `path: null` (matching prod), and a path leaf is `{ id, name }` (no `parent_id`).
- New `PATCH /aio/tags/{tag_id}` (`aio-update-tag`) re-parents/renames/promotes in place, returning **200** and **404 `{message:'not found'}`** for an unknown id (both prod-verified).
- The baked `workspace-with-data` seed ships a populated hierarchy (root `category:` → bare child) and the seeded prompt carries both tags.
- Spec overlay corrections from the live probe: CR11 (PATCH 200 not 201), CR12 (PATCH 404 for unknown id), CR13 (`AIOTag.parent_id`/`path` nullable). Regenerated types reflect these; no vendored-swagger edit.
- The "1-level" tree is documented as a product convention, not an API constraint — the probe showed prod accepts deeper nesting (a grandchild create returned 201).

Two edges whose live behaviour is still unconfirmed — deleting a parent that still has children (orphaning) and the mock's name-derived id making a post-rename re-create duplicate — remain documented as known limitations rather than resolved with invented behaviour.

## 4. Required information
- Jira / issue: https://github.com/adobe/spacecat-shared/issues/1756
- Spec: https://github.com/adobe/serenity-docs/issues/21
- Other: folds in https://github.com/adobe/spacecat-shared/issues/1754 (baked nested fixture) ; Postgres-seed counterpart https://github.com/adobe/mysticat-data-service/issues/757

## 5. Affected / used mysticat-workspace projects
- spacecat-api-service — consumer/contract: its serenity transport writes `parent_id` on create and reads the nested tag tree; this mock validates that path locally (tracking https://github.com/adobe/spacecat-api-service/issues/2733).
- mysticat-data-service — consumer/contract: the onboarding CLI writes nested tags via this API (https://github.com/adobe/mysticat-data-service/pull/737).
- The cross-repo e2e harness / workspace local-dev consume the mock (npm package + GHCR image) to exercise `/serenity/*` against a faithful tree.

## 6. Additional information outside the code
Live prod probe was RUN on 2026-07-01 against `adobe-hackathon.semrush.com`, test workspace `bb0f4e1c-…` (a child sub-workspace's live project) — the ACL gap noted in serenity-docs#21 §7 is resolved for the probing IMS user. Throwaway tags were created under a real category, inspected, and deleted (verified removed; `children_count` back to 0 — nothing left behind).

Findings:
- Confirmed matching the mock: create-child wire (`{names[], parent_id}` → 201 top-level array with `parent_id`), ancestors-only single-leaf `path`, `children_count`, `search` substring.
- Corrected to match prod (this PR): PATCH success is **200** not 201; PATCH unknown id is **404 `{message:'not found'}`** not 500; a flat root returns `parent_id: null` / `path: null` (not omitted / `[]`); a path leaf is `{ id, name }`; there is **no 1-level depth cap** on the wire.
- Observed but NOT modelled (surfaced separately): newly-created tags are visible only in the `?draft=true` view until the project is published; the live (non-draft) read hides them. The consumer's `listProjectTags` reads the live view with no `draft`, so a created category would not appear live until publish — worth confirming the consumer publishes before it reads categories back. Not changed here to avoid inverting the mock's deliberate "0-prompt category reads back immediately" contract without that confirmation.

## 7. Test plan
(a) Local end-to-end: booted the Counterfact mock and drove the real generated client through nested create (`POST parent_id`), child listing (`GET ?parent_id=` with `children_count` + `path`), PATCH re-parent/rename/promote, PATCH unknown-id → 404, delete-without-cascade, root `null` shape, and the baked-seed nested read. Plus the live prod probe above.
(b) Per-consumer verification (dev-tooling/mock package, not a deployed service): after publish, consumers bump the package / pull the refreshed GHCR mock image and run their serenity IT / local-dev against the nested tags — confirm a created child reads back under its parent, the Categories tree renders from the baked seed, and a re-parent reflects in the listing.

## 8. Deployment & merge order
Part of the cross-repo nested-category effort but independent to merge (additive, backward-compatible for flat callers):
- related: https://github.com/adobe/spacecat-api-service/issues/2733 — the proxy + create-tag work that writes `parent_id`.
- related: https://github.com/adobe/mysticat-data-service/pull/737 — the onboarding CLI that writes the nested tree.
- related: https://github.com/adobe/project-elmo-ui/pull/2078 — the Serenity Categories UI.

Recommended sequence: merge and publish this mock first so the consumer PRs can validate their nested behaviour locally against it; it does not depend on any of them.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/spacecat-shared-project-engine-client/mock/counterfact/routes/v2/workspaces/{id}/projects/{project_id}/aio/tags/{tag_id}.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  const scope = { workspaceId: path.id, projectId: path.project_id };
+  // Empty `parent_id` (or omitted) promotes the tag to a root — clear it to '' so the stored shape
+  // reflects "no parent" (the read path treats a falsy parent_id as a root).
+  const parentId = String(body?.parent_id ?? '');
+  const updated = context.ops.tags.update(scope, path.tag_id, {
+    name: body?.name,
+    parent_id: parentId,
+  });
+  if (!updated) {
+    return $.response[500].json(
```
Comment (MysticatBot) [inline packages/spacecat-shared-project-engine-client/mock/counterfact/routes/v2/workspaces/{id}/projects/{project_id}/aio/tags/{tag_id}.js]: issue (blocking): `body?.name` is `undefined` when the request body omits `name` (e.g. a re-parent-only PATCH with `{ parent_id: "new-parent" }`). This `undefined` is passed to `store.update` which spreads it onto the stored entity via `{ ...current, ...patch }`, setting `name: undefined`. Subsequent reads serialize a missing/undefined name, breaking consumers that expect a string.

Fix: fall back to the current stored name when `body.name` is absent. This requires either reading the entity before patching, or having `store.update` skip undefined values:

```js
const current = context.ops.tags.list(scope).find((t) => t.id === path.tag_id);
if (!current) { return $.response[500].json(...); }
const updated = context.ops.tags.update(scope, path.tag_id, {
  name: body?.name ?? current.name,
  parent_id: parentId,
});
```

File: packages/spacecat-shared-project-engine-client/test/e2e/project-engine-mock.e2e.js [hunk 2]
```diff
@@ -549,13 +549,14 @@ async function waitForReady(baseUrl, deadline, getStderr) {
   // The core fix: a standalone tag created via createProjectTags must PERSIST so a 0-prompt
   // category reads back via GET /aio/tags (today the elmo Categories surface falls back to an
   // optimistic add because the mock couldn't persist it). parent_id + search are spec-required
-  // query params (the consumer sends them); empty search lists every stored tag.
-  it('persists createProjectTags and reads the 0-prompt category back via GET /aio/tags', async () => {
+  // query params (the consumer sends them); empty parent_id lists the root categories. A brand-new
+  // name (distinct from the baked seed's `category:Running Shoes`) proves the create→read path.
```
Comment (MysticatBot) [inline packages/spacecat-shared-project-engine-client/test/e2e/project-engine-mock.e2e.js:557]: issue (blocking): The PATCH E2E test exercises promote-to-root but never exercises a rename (`body: { name: "NewName", parent_id: ... }`). The route handler builds the response from `updated.name` - if `store.update` were to return the pre-patch entity (or the response construction had a typo), only an E2E rename assertion would catch it. The unit test in `stateful.test.js` covers the store layer, but the full route-handler-to-response roundtrip for rename is unproven.

Fix: add one assertion in the existing PATCH test (or a brief new case) that PATCHes with a new name and asserts both the response and a subsequent GET reflect the updated name.

**Review by MysticatBot — CHANGES_REQUESTED** [review]
Hey @rainer-friederich,

**Verdict: Request changes** - one behavioral bug in the PATCH handler that corrupts stored state when `name` is omitted from the request body.
**Complexity: HIGH** - large diff; API surface signal.
_Changes: adds 1-level nested AIO tag support to the Counterfact mock - parent-aware create, tree-aware listing, PATCH re-parent/rename, and a nested baked seed (14 files)._

## Must fix before merge

1. [Important] PATCH with missing `name` writes `undefined` into the stored entity - `mock/counterfact/routes/v2/workspaces/{id}/projects/{project_id}/aio/tags/{tag_id}.js:42` (details inline)
2. [Important] PATCH rename path has no E2E coverage - `test/e2e/project-engine-mock.e2e.js` (details inline)

<details>
<summary>Non-blocking (4): minor issues and suggestions</summary>

- nit: GET handler computes `children_count` with O(n^2) scan inside `.map()` - a pre-computed `Map<parentId, count>` would be O(n) and clearer in intent - `mock/counterfact/routes/.../aio/tags.js:139`
- suggestion: PATCH returns 500 for a not-found tag; 404 is the HTTP-standard default when live behavior is unknown and avoids misleading future debuggers - `mock/counterfact/routes/.../aio/tags/{tag_id}.js:47`
- nit: E2E nested-create test does not assert `children_count: 0` on the newly created child leaf - `test/e2e/project-engine-mock.e2e.js:500`
- suggestion: add a one-line comment at the POST `upsertMany` call site noting that re-POSTing an existing tag with a different `parent_id` is a no-op (returns stored version) to prevent a future contributor from filing a bug - `mock/counterfact/routes/.../aio/tags.js:63`

</details>

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 2m 3s | <b>Cost:</b> $5.91 | <b>Commit:</b> <code>4a9350d7908de06683d58ede4b7267d46705d5d1</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

File: packages/spacecat-shared-project-engine-client/mock/counterfact/routes/v2/workspaces/{id}/projects/{project_id}/aio/tags/{tag_id}.js (hunk 1 above)
Comment (rainer-friederich) [reply packages/spacecat-shared-project-engine-client/mock/counterfact/routes/v2/workspaces/{id}/projects/{project_id}/aio/tags/{tag_id}.js]: Declined: `name` is required in `model.TreeNodeRequest` and Counterfact request validation is ON (`mock/run.js`), so a PATCH without `name` 400s before the handler runs — the `undefined` never reaches `store.update`. A stored-name fallback would be dead code (coverage-excluded handler, only reachable with validation off). See the consolidated comment.

File: packages/spacecat-shared-project-engine-client/test/e2e/project-engine-mock.e2e.js (hunk 2 above)
Comment (rainer-friederich) [reply packages/spacecat-shared-project-engine-client/test/e2e/project-engine-mock.e2e.js:557]: Fixed in `f1cd3aa2`: added a PATCH rename roundtrip E2E case — PATCH `{ name: "Hiking", parent_id }` and assert both the 201 response and a subsequent GET reflect the new name.

**Review by MysticatBot — APPROVED** [review]
Hey @rainer-friederich,

**Verdict: Approve** - clean re-review; both prior blocking findings resolved, no new blockers.
**Complexity: HIGH** - large diff; API surface signal.
_Changes: adds 1-level nested AIO tag support to the Counterfact mock - parent-aware create, tree-aware listing, PATCH re-parent/rename, and a nested baked seed (14 files)._

<details>
<summary>Non-blocking (3): minor issues and suggestions</summary>

- suggestion: PATCH returns 500 for a not-found tag; when the live behaviour is confirmed, 404 would be the conventional default and avoids misleading consumers who develop against the mock - `mock/counterfact/routes/.../aio/tags/{tag_id}.js:47`
- nit: a self-referential PATCH (parent_id equal to the tag's own id) is silently accepted and produces a confusing path breadcrumb pointing at itself; consider a code comment acknowledging the edge - `mock/counterfact/routes/.../aio/tags/{tag_id}.js:41`
- suggestion: the DELETE-orphan scenario (parent deleted, child becomes invisible in both roots and children listings) is documented in the header but has no test anchoring the behaviour for consumers - a single E2E assertion proving the orphan state would make the documented limitation concrete - `test/e2e/project-engine-mock.e2e.js`

</details>

## Previously flagged, now resolved
- PATCH rename path now has E2E coverage (f1cd3aa2)
- GET children_count uses pre-computed O(n) Map (no longer O(n^2))
- POST idempotent re-parent no-op documented with inline comment
- New child leaf asserts children_count: 0

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 1m 10s | <b>Cost:</b> $7.26 | <b>Commit:</b> <code>f1cd3aa278c9ab93a16573da6fbc3c9ba2209322</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by MysticatBot — APPROVED** [review]
Hey @rainer-friederich,

**Verdict: Approve** - clean re-review; incremental commit addresses both prior suggestions, no new concerns.
**Complexity: HIGH** - large diff; API surface signal.
_Changes: adds self-referential PATCH comment and DELETE-orphan E2E test addressing prior review suggestions (2 files)._

<details>
<summary>Non-blocking (0): no issues</summary>

No findings.
</details>

## Previously flagged, now resolved
- Self-referential PATCH edge case now documented with inline comment (9700f108)
- DELETE-orphan limitation now anchored by E2E test proving the child drops out of roots and survives only via the stale parent id (9700f108)
- PATCH 500 vs 404 for not-found tag: author rationale accepted (response validation ON, 404 undeclared in spec)

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 0m 43s | <b>Cost:</b> $3.85 | <b>Commit:</b> <code>9700f108c67cb26e87190f1ec7f994d2c627cb3c</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by MysticatBot — APPROVED** [review]
Hey @rainer-friederich,

**Verdict: Approve** - clean re-review; the live-probe corrections (CR11/CR12/CR13) are coherent across overlay, types, mock, and tests.
**Complexity: HIGH** - large diff; API surface signal.
_Changes: reconciles the Counterfact mock and spec overlay with live prod probe findings - PATCH 200/404, nullable parent_id/path, path leaf shape (16 files)._

<details>
<summary>Non-blocking (3): minor issues and suggestions</summary>

- nit: PATCH handler JSDoc says "-> 201 TreeNodeResponse" but implementation returns 200 (the block comment above it correctly says 200) - `mock/counterfact/routes/.../aio/tags/{tag_id}.js:31`
- nit: GET handler header comment says "a root's `path` is `[]`" but implementation now returns `null` (CR13) - `mock/counterfact/routes/.../aio/tags.js:67`
- nit: E2E rename test comment says "reflected in the 201 response" but PATCH now returns 200 - `test/e2e/project-engine-mock.e2e.js:721`

</details>

## Previously flagged, now resolved
- PATCH 500 vs 404 for not-found tag: now returns 404 with CR12 overlay declaring it (1f6888bf)
- GET children_count uses pre-computed O(n) Map
- PATCH rename has E2E coverage (f1cd3aa2)
- Self-referential PATCH documented with inline comment (9700f108)
- DELETE-orphan limitation anchored by E2E test (9700f108)
- New child leaf asserts children_count: 0
- POST idempotent re-parent no-op documented

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 0m 59s | <b>Cost:</b> $3.93 | <b>Commit:</b> <code>1f6888bfa9bc1caee39eca6c360add8650f7a842</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by MysticatBot — APPROVED** [review]
Hey @rainer-friederich,

**Verdict: Approve** - all prior nits resolved; no new concerns.
**Complexity: HIGH** - large diff; API surface signal.
_Changes: fixes 3 stale-comment nits from prior review - JSDoc and header comments now match the CR11/CR12/CR13 prod-probe corrections (3 files)._

## Previously flagged, now resolved
- PATCH JSDoc "-> 201" corrected to "-> 200" (8e250d87)
- GET header comment "a root's `path` is `[]`" corrected to "`parent_id` and `path` are `null`" (8e250d87)
- E2E rename test comment "201 response" corrected to "200 response" (8e250d87)

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 5m 47s | <b>Cost:</b> $1.43 | <b>Commit:</b> <code>8e250d874b7726b4937b9d7437076e82734f8aca</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by MysticatBot — APPROVED** [review]
Hey @rainer-friederich,

**Verdict: Approve** - clean re-review; merge commits from main integrate cleanly with no new concerns.
**Complexity: HIGH** - large diff; API surface signal.
_Changes: merges main into the feature branch, integrating the sub-workspace seed refactor alongside the nested-tag work (16 files)._

<details>
<summary>Non-blocking (2): suggestions</summary>

- suggestion: E2E PATCH tests (promote then rename) share mutated state sequentially; a `__reset` between them would make them independent and order-agnostic - `test/e2e/project-engine-mock.e2e.js`
- suggestion: the `two-hierarchies` seed uses flat tags for H2 (`aioTags(...)`) so no nested-tag coverage from the second hierarchy; fine for now but worth noting if a consumer exercises Categories against H2 - `mock/seeds.js`

</details>

## Previously flagged, now resolved
- PATCH JSDoc "-> 201" corrected to "-> 200" (8e250d87)
- GET header comment "path is []" corrected to "parent_id and path are null" (8e250d87)
- E2E rename test comment "201 response" corrected to "200 response" (8e250d87)
- Self-referential PATCH edge case documented (9700f108)
- DELETE-orphan limitation anchored by E2E test (9700f108)
- PATCH 500 vs 404 for not-found tag: now returns 404 with CR12 overlay (1f6888bf)

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 0m 51s | <b>Cost:</b> $6.72 | <b>Commit:</b> <code>73fd57f817e1a7c9e25e870033c065f16e3eb405</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>
